apps/dcos_dns/src/dcos_dns.erl

```diff
@@ -189,12 +198,58 @@ push_prepared_zone(ZoneName, Records) ->
             end
     catch error:Error ->
         lager:error(
-            "Failed to push DNS Zone \"~s\": ~p",
-            [ZoneName, Error]),
+            "Failed to push DNS Zone \"~s\": ~p", [ZoneName, Error]),
         {error, Error}
     end.
 
+%%%===================================================================
+%%% Internal functions
+%%%===================================================================
+
 -spec(bin_to_hex(binary()) -> binary()).
 bin_to_hex(Bin) ->
     Bin0 = << <<(integer_to_binary(N, 16))/binary>> || <<N:4>> <= Bin >>,
     cowboy_bstr:to_lower(Bin0).
+
+-spec(emit_metrics(binary(), integer(), [dns:dns_rr()]) -> ok).
```

Comment:
As we discussed earlier, please don't create metric functions when it's possible. In this case, I'd recommend to add `put_zone/2` function that will push data to erldns cache and do all logging and metric stuff, i.e. something like this (I didn't test it):

```erlang
-spec(push_prepared_zone(dns:dname(), [dns:dns_rr()]) ->
    ok | {error, Reason :: term()}).
push_prepared_zone(ZoneName, Records) ->
    Begin = erlang:monotonic_time(),
    try
        case erldns_zone_cache:get_zone_with_records(ZoneName) of
            {ok, #zone{records=Records}} ->
                ok;
            _Other ->
                put_zone(Begin, ZoneName, Records)
        end
    catch error:Error ->
        lager:error("Failed to push DNS Zone ~s: ~p", [ZoneName, Error]),
        {error, Error}
    end.

%%%===================================================================
%%% Internal functions
%%%===================================================================

-spec(put_zone(Begin :: integer(), dns:dname(), [dns:dns_rr()]) -> ok).
put_zone(Begin, ZoneName, Records) ->
    Hash = crypto:hash(sha, term_to_binary(Records)),
    case erldns_zone_cache:put_zone({ZoneName, Hash, Records}) of
        ok ->
            Duration = erlang:monotonic_time() - Begin,
            ZoneName0 = <<ZoneName/binary, ".">>,
            lager:notice(
                "DNS Zone ~s was updated (~p records, sha: ~s, duration: ~pms)",
                [ZoneName0, length(Records), bin_to_hex(Hash),
                 erlang:convert_time_unit(Duration, native, millisecond)]),
            prometheus_gauge:set(
                dns, zone_records,
                [ZoneName0], length(Records)),
            prometheus_summary:observe(
                dns, zone_push_duration_seconds,
                [ZoneName0], Duration);
        {error, Error} ->
            error(Error)
    end.
```